### ISSUE TITLE
Addition of anonymous block initialization analysis in Sec 3.8 of checker manual

### ISSUE THREAD

**Issue by ashishrana160796** [typetools/checker-framework#1837]
Considering the following case, class `JsonWriter` from `gson` library uses anonymous block for calling a method `push` while class object is still in `UnderInitialization` state. The issue here is annotating around such particular condition. This snippet part 
```
{
    push(EMPTY_DOCUMENT);
 }
```  
 is producing the following error :
```
 error: [method.invocation.invalid] call to push(int) not allowed on the given receiver.
    push(EMPTY_DOCUMENT);
        ^
  found   : @UnderInitialization(java.lang.Object.class) @NonNull JsonWriter
  required: @UnknownInitialization(com.google.gson.stream.JsonWriter.class) @NonNull JsonWriter
``` 
By using sections 3.8(_Introduction_),3.8.6(_handling warnings_) from manual another way of altering the code, in the following manner does the task but this approach is not working for anonymous block. Here is the altered snippet.
```
public JsonWriter(Writer out) {
    if (out == null) {
      throw new NullPointerException("out == null");
    }
    this.out = out;
    push(EMPTY_DOCUMENT);    // Anonymous Block removed
  }
```
```
private void push(/*>>>@UnknownInitialization(JsonWriter.class) JsonWriter this,*/ int newTop) {
...
}
```
This produces no error and works in a similar fashion but here class instance type has also changed after constructor execution to `@UnderInitialization(com.google.gson.stream.JsonWriter.class) @NonNull JsonWriter`. But this approach fails to catch the bug. This is related to the analysis missing from the checker manual. By the following snippet we can annotate anonymous block when the type of underinitialized instance of type `java.lang.Object` .
`private void push(/*>>>@UnknownInitialization(java.lang.Object.class) JsonWriter this,*/ int newTop) {`

The issue here is annotation of such blocks as the class code( that can be buggy like this one). If possible an addition of handling such case in `checker manual` where `receiver` methods can also get called from anonymous block would be really helpful in future references. 

Annotated File : 
[JsonWriter.txt](https://github.com/typetools/checker-framework/files/1725687/JsonWriter.txt)

**Comment by wmdietl** [comment]
Can you please follow the instructions at https://checkerframework.org/manual/#reporting-bugs
and try to report a minimal, self-contained test case for your issue?

**Comment by ashishrana160796** [comment]
The main issue is highlighted in following code :
```
class A
{
	int[] a = new int[32];	
	{
		rec();
	}
	A()
	{
		//rec();
	}	
	void rec(/*>>>@UnknownInitialization A this*/)
	{
		if(a.length != 0)
		{
			int[] b = new int[32];
		}
	}
}
```
Here in above code `a` is initialized beforehand when anonymous block(or class constructor) is calling th e `rec` method. But still an error regarding possible dereference is coming. Which shouldn't come as it already initialized to a non-null type at the moment it is created.
Error :
```
error: [dereference.of.nullable] dereference of possibly-null reference a
		if(a.length != 0)
```

**Comment by wmdietl** [comment]
The Checker Framework performs a modular analysis and only looks at the specifications of other fields/methods.
In your example method `rec` uses an `@UnknownInitialization A` receiver parameter `this`. This expresses that the method can be called on an `A` object that has been initialized only up to the `java.lang.Object` frame. Therefore, field `a` has to be considered `@Nullable` and you get the right error message for the example.

If you change the code to use the right framing, it type checks:

```java
import org.checkerframework.checker.initialization.qual.UnderInitialization;

class A {
    int[] a = new int[32];
	
    {
        rec();
    }

    void rec(@UnderInitialization(A.class) A this) {
        if (a.length != 0) {
            int[] b = new int[32];
        }
    }
}
```

The Initialization Checker manual section is at https://checkerframework.org/manual/#initialization-checker
Please let me know if this clarifies what is happening here. This behavior is as intended.
